#15

Hanka8 opened the pull request:
#15 alternative solution

Hanka8 committed:
test

Hanka8 committed:
Add no birds message

Hanka8 committed:
feat: add side panel with bird details

Add popup display bird details when a bird marker is clicked on the map.
Move wikidata to Index component and pass it down to BirdCard and BirdPopup components.

Hanka8 committed:
feat: Implement side panel with interactive map and bird popup components

Edit FetchBirdData function to include more cases
Add fallback logic to wikiData in Birdpopup.tsx

Hanka8 committed:
feat: implement side panel with map and bird details

Center map and animate marker when user clicks on birds location.

tkudlicka commented on package.json:
```diff
@@ -11,12 +11,16 @@
+    "@types/leaflet": "^1.9.15",
     "date-fns": "^3.6.0",
+    "leaflet": "^1.9.4",
     "ol": "^10.2.1",
```
remove prev map library
```suggestion
```

tkudlicka commented on src/components/BirdPopup.tsx:
```diff
@@ -0,0 +1,119 @@
+
+  const currentBird = birds[currentIndex];
+  
+  const wikiData =
+    wikiDataMap.get(currentBird.comName.toUpperCase()) ||
+    // if no bird change eurasian to common and vice versa
+    wikiDataMap.get(
+      currentBird.comName
+        .replace("EURASIAN", "COMMON")
+        .replace("COMMON", "EURASIAN")
+        .toUpperCase()
+    ) ||
+    // if rook go to bird page
+    (currentBird.comName.toUpperCase() === "ROOK"
+      ? wikiDataMap.get("ROOK (BIRD)")
+      : null) ||
+    // if rock pigeon go to rock dove
+    (currentBird.comName.toUpperCase() === "ROCK PIGEON"
+      ? wikiDataMap.get("ROCK DOVE")
+      : null) ||
+      // replace "-" with " "
+      (currentBird.comName.includes("-")
+        ? wikiDataMap.get(currentBird.comName.replace(/-/g, " ").toUpperCase())
+        : null) ||
+    wikiDataMap.get(
+      currentBird.comName.split(" ").slice(1).join(" ").toUpperCase()
+    ) ||
+    Array.from(wikiDataMap.entries()).find(([key]) =>
+      key.includes(
+        currentBird.comName.split(" ").slice(1).join(" ").toUpperCase()
+      )
+    )?.[1];
```
The current implementation has redundant computations (e.g., repeated .toUpperCase() and .replace() calls), making it inefficient and harder to maintain. It lacks preprocessing for currentBird.comName, leading to unnecessary recalculations. The nested logic is complex and less readable, increasing the risk of errors when adding or modifying transformations.

Hanka8 replied:
Commit 54a929c

tkudlicka commented on src/components/Index.tsx:
```diff
@@ -78,6 +80,52 @@ const Index: React.FC = () => {
+      );
+      if (response.ok) {
+        const data = await response.json();
+        if (data.type !== "disambiguation") {
+          return data;
+        }
```
this function is overly complex as it tries to handle multiple responsibilities. It should be simplified to focus solely on fetching data and returning it, without making multiple calls to different endpoints or performing unrelated logic.

Hanka8 replied:
Commit 053992f

tkudlicka commented on src/components/Index.tsx:
```diff
@@ -78,6 +80,52 @@ const Index: React.FC = () => {
+      }
+
+    } catch (error) {
+      console.error("Error fetching bird data from Wikipedia:", error);
+      return {};
```
it should probably throw error and if there is error than show error boundary component or alert, it couldnt download bird wiki data

Hanka8 replied:
minor change in Commit 733a64b

I think that wikierror is actually passes as a prop to BirdCard

tkudlicka commented on src/components/Index.tsx:
```diff
@@ -89,6 +137,28 @@ const Index: React.FC = () => {
     queryFn: () => fetchBirdsNearby({ latitude, longitude }),
   });
 
+  // Add Wikipedia data queries
+  const wikiQueries =
+    data?.map((bird) => ({
+      queryKey: ["birdData", bird.comName],
+      queryFn: () => fetchBirdData(bird.comName),
+    })) || [];
+
+  const wikiResults = useQueries({
+    queries: wikiQueries,
+  });
+  
+  // Create a map for faster lookups
+  const wikiDataMap = useMemo(() => {
+    const map = new Map();
+    wikiResults.forEach((result) => {
+      if (result.data?.title) {
+        map.set(result.data.title.toUpperCase(), result.data);
+      }
+    });
+    return map;
+  }, [wikiResults]);
+
```
I think this is unnecessary

Hanka8 replied:
I dont get this one because I am using wikiDataMap in BirdPopup.tsx for pairing bird photo with its name - bird popups would not show bird images without it. Memoization seems useful there as it prevents unnecessary recreations of the map and wikiQueries are there to fetch data for all the birds. Can you be more specific? 🤔🙂

tkudlicka reviewed (changes requested):
Great work on this implementation! It’s well-structured and clear overall. However, there are a few areas that could benefit from improvement to enhance readability and efficiency.

Hanka8 committed:
refactor(BirdPopup): bird name transformations

- Reduce redundant string operations
- Add sequential transformation attempts
- Add fallback strategy with partial match
- Improve maintainability and readability

Hanka8 committed:
Update package.json

Co-authored-by: Tomáš Kudlička <70945365+tkudlicka@users.noreply.github.com>

Hanka8 committed:
Merge branch 'hm/side-panel' of github.com:Hanka8/BirdFinder into hm/side-panel

Hanka8 committed:
refactoring: extract pure fetch function for Wiki

- Move transformations outside function scope
- Split Wikipedia data fetching into separate pure function
- Reduce function complexity and responsibilities

Hanka8 committed:
refactoring: update Index.tsx error handling

- fetchBirdData function now throws errors

tkudlicka reviewed (approved):
🔥 lgtm
